### DISCUSSION TITLE
Authorization Guard.

### DISCUSSION THREAD

**Discussion by hanza93** [urql-graphql/urql#2391] [Q&A] [answered]
Hello, I'm trying to create an authorization guard with some permissions and I don't know exactly how to manage this.
I made the UrqlClientOptions as function 
```javascript
export const getUrqlClientOptions =
  (url: string, withAuth: boolean) =>
  (ssrCache: SSRExchange = ssrExchange({ isClient: false })) => ({
    url,
    requestPolicy: 'cache-and-network',
    exchanges: [
      devtoolsExchange,
      dedupExchange,
      ...(withAuth
        ? [
            authExchange({
              getAuth,
              addAuthToOperation,
              didAuthError,
              willAuthError,
            }),
          ]
        : []),
      // cacheExchange({}),
      ssrCache,
      fetchExchange,
      multipartFetchExchange,
    ],
  });
```
to create `withAuth` and `withoutAuth`

and my `getAuth` function is a pretty standard
```javascript
const getAuth = async ({ authState }) => {
  if (!authState) {
    const authStore = getAuthStore();
    const accessToken = authStore?.accessToken;

    if (accessToken) {
      return { ...authStore };
    }

    return null;
  }

  destroyAuthStore();
  Router.push('/auth/sign-in');

  return null;
};
```
and by using `withAuth` or `withoutAuth` as HOC on everypage, I can tell if this is a private or public page.
what I couldn't figure out is
1. redirect on login/register pages if the user already logged in.
2. let's say some of pages needs the user to be with role`XYZ`.

**Comment by JoviDeCroock** [comment, marked as answer]
The redirect issue can be solved by checking whether there is a token and redirecting if there is.

There's a few ways to go about this, generally the one I take is to react to errors so basically when you are performing queries you can look at what error you are getting and when it's `Forbidden` or `Unauthorized` you know people have no permissions for a given page.

The second thing you can do is sync up your `authState` from your client to your UI through an event-emitter like [mitt](https://github.com/developit/mitt) and check synchronously in your UI.
